index.js

```diff
@@ -498,37 +537,49 @@ MediaSession.prototype = extend(MediaSession.prototype, {
-        if (desc.contents.length > 0) {
-            this.send('source-remove', desc);
-        }
+
+    // Justin's new functions
+    _determineDifferencesAndSignal: function(oldLocalDescription, newLocalDescription) {
+
+
+        const oldDescCopy = JSON.parse(JSON.stringify(oldLocalDescription));
+        const newDescCopy =  JSON.parse(JSON.stringify(newLocalDescription));
+
+        var diffObject = generateDifferenceOfSources(oldDescCopy, newDescCopy);
+
+        this._signalDifferenceiInSources({sourcesRemoved: diffObject.sourcesRemoved, sourcesAdded: diffObject.sourcesAdded,  sourcesModified: diffObject.sourcesModified,
+             oldLocalDescription: oldDescCopy, newLocalDescription: newDescCopy });
     },
 
-    _addRecvOnlySourceIfNotPresent: function(oldLocalDescription, newLocalDescription) {
-        var desc = newLocalDescription;
+    _signalDifferenceiInSources: function(diffObject){
+        const oldContents = diffObject.oldLocalDescription.contents.filter(getContent);
+        const newContents = diffObject.newLocalDescription.contents.filter(getContent);
+        delete diffObject.oldLocalDescription.groups;
+        delete diffObject.newLocalDescription.groups;
 
-        // filter to only sources that changed from recvonly to sendrecv
-        desc.contents = desc.contents.filter(function(newContent) {
-            if (newContent.application.applicationType === 'rtp' && newContent.application.sources && newContent.application.sources.length) {
-                var oldContent = findMatchingContentBlock(newContent, oldLocalDescription);
-                // the filter function handles the case where oldContent is null
-                return filterToMatchingRecvonly(newContent, oldContent);
-            }
-        });
-        delete desc.groups;
-        if (desc.contents.length > 0) {
-            this.send('source-add', desc);
+        if (diffObject.sourcesAdded.length) {
+            diffObject.newLocalDescription.contents = getProperSSRCS(newContents, diffObject.sourcesAdded);
+            this._log('info', 'sending source add', diffObject.newLocalDescription);
+            this.send('source-add', diffObject.newLocalDescription);
+        }
+
+        if (diffObject.sourcesRemoved.length && !diffObject.sourcesModified.length) { // to avoid signaling remove twice
+            diffObject.oldLocalDescription.contents = getProperSSRCS(oldContents, diffObject.sourcesRemoved);
+            this._log('info', 'sending source remove', diffObject.oldLocalDescription);
+            this.send('source-remove', diffObject.oldLocalDescription);
+        }
+
+        if (diffObject.sourcesModified.length) {
+            diffObject.oldLocalDescription.contents = getProperSSRCS(oldContents, diffObject.sourcesModified);
+            diffObject.newLocalDescription.contents =  getProperSSRCS(newContents, diffObject.sourcesModified);;
+            const filteredDesc = filterMsidFromRecvonlySources(diffObject.oldLocalDescription);
```

Comment:
Is there a case where we'd need to apply this filter for added or removed sources? Maybe getProperSSRCS could do this filter internally?

Reply:
Thinking about it i'm having a hard time thinking of a case where we would need to filter out the msid from just a source add or remove. all the cases where it matter should be in the sources modified block. i don't think it would matter to add that filter to the other cases but i'm inclined to just call it on the description again rather than having `getProperSSRCS` modify it.
Looking into it more we don't have the direction to filter on in the get Proper so i'm inclined to filter it in a separate call if you think it's needed.